tools/istio-iptables/pkg/cmd/run.go

```diff
@@ -467,10 +490,22 @@ func (iptConfigurator *IptablesConfigurator) run() {
 	}
 
 	if redirectDNS {
-		// redirect all TCP dns traffic on port 53 to the agent on port 15053
-		iptConfigurator.iptables.AppendRuleV4(
-			constants.ISTIOOUTPUT, constants.NAT, "-p", constants.TCP, "--dport", "53", "-j", constants.REDIRECT, "--to-ports", constants.IstioAgentDNSListenerPort)
-		// the rest of the IPtables rules will take care of ensuring that the traffic does not loop, among other things.
+		for _, s := range iptConfigurator.cfg.DNSServersV4 {
+			// redirect all TCP dns traffic on port 53 to the agent on port 15053 for all servers
+			// in etc/resolv.conf
+			// We avoid redirecting all IP ranges to avoid infinite loops when there are local DNS proxies
+			// such as: app -> istio dns server -> dnsmasq -> upstream
+			// This ensures that we do not get requests from dnsmasq sent back to the agent dns server in a loop.
+			// Note: If a user somehow configured etc/resolv.conf to point to dnsmasq and server X, and dnsmasq also
+			// pointed to server X, this would not work. However, the assumption is that is not a common case.
```

Comment:
Yeah, usually user don't put the root name servers in resolv.conf while dnsmasq is likely to contain.